### PR TITLE
PF-7364

### LINKED ISSUE #7364: FileUpload: Multiple with auto="true" only works first time
**Describe the defect**
Getting irregular and unreliable results when performing an auto, advanced, multiple file upload.  
On startup you simply get a file upload instance to select csv files from. I'm assuming you have, or can get a set of csv files (~4) to be able to use for this. On first selection it appears to work correctly, i.e., select 4 files and upon submit, the table showing files selected will list all four.  
The `Clear List` button does just that... it removes the file selected list and refreshes the section and should see no files selected.  
Now do this again... select the same four files. I'm seeing that subsequent file selections are being missed... in fact, most of the time.

**Reproducer**
I have a fork of the primefaces-test repo [here.](https://github.com/karlkras/primefaces-test)

**Environment:**
 - PF Version: _10.0_
 - JSF + version: _e.g. Mojarra 2.2.20 or MyFaces 2.3.9_
 - Affected browsers: _e.g. Chrome, iOS, ALL_

**To Reproduce**
1. Have 4 csv files available in a test directory.
2. Click on the `Choose` button (FileUpload control)
3. Navigate to directory with test csv files, select all.
4. When updated, you will see list of files selected in a table below the control.
5. (First time in, it appear to show all files correctly.)
6. Click on `Clear List`
7. File resources will be deleted from backing bean, form updated, and fileupload will be available again.
8. Repeat #s 2-4 
9. Subsequent selections will show irregular results in the files that were selected. Mostly incorrectly.

**Expected behavior**
I would expect to see the resulting file list represent exactly what I had selected.

```

### PR DESCRIPTION
**Pull request by wal-jan**
As part of my analysis on FileUpload, I created some integration tests.

### CODE REVIEW

File: primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/fileupload/AbstractFileUploadTest.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.stream.Collectors;
+import org.junit.jupiter.api.Assertions;
+import org.primefaces.selenium.AbstractPrimePageTest;
+import org.primefaces.selenium.component.DataTable;
+import org.primefaces.selenium.component.model.datatable.Row;
+
+public abstract class AbstractFileUploadTest extends AbstractPrimePageTest {
+
+    protected File locateClientSideFile(String fileName) {
+        String folder = AbstractFileUploadTest.class.getPackageName().replace(".", "/");
```
Comment (christophs78) [inline primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/fileupload/AbstractFileUploadTest.java]: Maybe as simple as `getPackage().getName()` to make Java8 happy.

**Comment by christophs78** [comment]
Thanks for your work! This is something we had been missing before. Happy you filled this gap!

**Comment by wal-jan** [comment]
Made it JDK 8 compatible. But the tests fail in MyFaces/Chrome combination. I'll investigate...

File: primefaces-selenium/primefaces-selenium-components/src/main/java/org/primefaces/selenium/component/FileUpload.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+    /**
+     * Waits until all selected files are uploaded.
+     * This only works in advanced mode.
+     */
+    public void waitAdvancedUntilAllFilesAreUploaded() {
+        ConfigProvider config = ConfigProvider.getInstance();
+        WebDriver driver = WebDriverProvider.get();
+
+        // todo: Should there be a separate configuration entry for file upload?
```
Comment (melloware) [inline primefaces-selenium/primefaces-selenium-components/src/main/java/org/primefaces/selenium/component/FileUpload.java:117]: Why not add this code to `PrimeExpectedConditions` class instead?

**Review by melloware — CHANGES_REQUESTED** [review]

**Comment by melloware** [comment]
Great work.  This was definitely a huge gap in our tests!

File: primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/fileupload/AbstractFileUploadTest.java [hunk 3]
```diff
@@ -29,12 +29,23 @@
...
 
 public abstract class AbstractFileUploadTest extends AbstractPrimePageTest {
+    /**
+     * The file upload tests only work in conjunction with Mojarra.
+     * Unfortunately MyFaces does not work.
+     * Call {@code assumeMojarra()} at the beginning of each file upload test.
+     * This method should be removed if file upload tests also work with MyFaces.
+     */
+    protected void assumeMojarra() {
+        Assumptions.assumeTrue(!FacesContext.class.getPackage().getImplementationVendor().contains("Apache"));
```
Comment (melloware) [inline primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/fileupload/AbstractFileUploadTest.java]: @christophs78 there might be a better way to do this like your TestTag that only runs if Safari etc?  That way you can mark the whole class.

@tandraschko we should get to the bottom of why these tests don't work with MyFaces.

**Review by melloware — CHANGES_REQUESTED** [review]

File: primefaces-selenium/primefaces-selenium-components/src/main/java/org/primefaces/selenium/component/FileUpload.java (hunk 2 above)
Comment (wal-jan) [reply primefaces-selenium/primefaces-selenium-components/src/main/java/org/primefaces/selenium/component/FileUpload.java:117]: I looked for a method that waits until all Javascript activity on client side is completed. But I did not find such a method. This method `waitAdvancedUntilAllFilesAreUploaded` only works for file upload widgets in advanced mode. From my point of view I would not move this code into `PrimeExpectedConditions`. But maybe there is a global solution to wait until all client side activity is completed, something like `PrimeSelenium#guardAjax()`.

File: primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/fileupload/AbstractFileUploadTest.java (hunk 3 above)
Comment (wal-jan) [reply primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/fileupload/AbstractFileUploadTest.java]: I found [p:FileUPload doesn't work Primefaces + Quarkus + MyFaces](https://forum.primefaces.org/viewtopic.php?t=62002) discussion. So I added context parameter primefaces.UPLOADER = commons to web.xml (and `commons-fileupload` dependency). But no luck, it didn't work. Afterwards I decided to skip these tests in conjunction with MyFaces.

**Comment by christophs78** [comment]
I checked out your PR and did some testing on my machine. (Windows 10 + up2date Firefox and Chrome)
The issue is not related to MyFaces or Mojarra. It´s related to Chrome.
Chrome fails with MyFaces and Mojarra.
Firefox work wiht MyFaces and Mojarra.
 
does not matter:
image

does matter:
image

**org.openqa.selenium.InvalidArgumentException: invalid argument**
  (Session info: chrome=91.0.4472.124)
  (Driver info: chromedriver=91.0.4472.101 (af52a90bf87030dd1523486a1cd3ae25c5d76c9b-refs/branch-heads/4472@{#1462}),platform=Windows NT 10.0.19043 x86_64) (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 0 milliseconds
Build info: version: '3.141.59', revision: 'e82be7d358', time: '2018-11-14T08:17:03'
System info: host: 'DESKTOP-6I7IEPJ', ip: '192.168.0.7', os.name: 'Windows 10', os.arch: 'amd64', os.version: '10.0', java.version: '1.8.0_292'
Driver info: org.openqa.selenium.chrome.ChromeDriver
Capabilities {acceptInsecureCerts: false, acceptSslCerts: false, applicationCacheEnabled: false, browserConnectionEnabled: false, browserName: chrome, chrome: {chromedriverVersion: 91.0.4472.101 (af52a90bf870..., userDataDir: C:\Users\chris\AppData\Loca...}, cssSelectorsEnabled: true, databaseEnabled: false, goog:chromeOptions: {debuggerAddress: localhost:60806}, handlesAlerts: true, hasTouchScreen: false, javascriptEnabled: true, locationContextEnabled: true, mobileEmulationEnabled: false, nativeEvents: true, networkConnectionEnabled: false, pageLoadStrategy: normal, platform: WINDOWS, platformName: WINDOWS, proxy: Proxy(), rotatable: false, setWindowRect: true, strictFileInteractability: false, takesHeapSnapshot: true, takesScreenshot: true, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unexpectedAlertBehaviour: ignore, unhandledPromptBehavior: ignore, version: 91.0.4472.124, webStorageEnabled: true, webauthn:extension:largeBlob: true, webauthn:virtualAuthenticators: true}

...
	at org.primefaces.selenium.internal.proxy.ElementLocatorInterceptor.intercept(ElementLocatorInterceptor.java:69)
	at org.primefaces.selenium.component.FileUpload$ByteBuddy$iu5H0Rcc.click(Unknown Source)
	**at org.primefaces.selenium.component.base.ComponentUtils.sendKeys(ComponentUtils.java:101)
	at org.primefaces.selenium.component.FileUpload.setValue(FileUpload.java:72)
	at org.primefaces.integrationtests.fileupload.FileUpload000Test.testBasicSingleUpload(FileUpload000Test.java:51)**

My current personal feeling is, that Chrome may not like sendKey on a file-input. Or we need to do this in a other way on Chrome.

**Comment by christophs78** [comment]
Some more information for the case it´s not self-explaining:

- Our regular CI runs integration-tests only for Java 11 + Chrome + Myfaces. (https://github.com/primefaces/primefaces/blob/master/.github/workflows/build.yml#L38)
- Additional we run once a night our Integrationtests against a extensive matrix. (https://github.com/primefaces/primefaces/blob/master/.github/workflows/nightly.yml)

**Comment by wal-jan** [comment]
@christophs78 thanks for your analysis! I'll try to fix it.
(Selenium does not support the OS dependent file selection dialogs out of the box. But I found multiple hints at Stackoverflow just to use sendKeys on input type=file. At least for Firefox and Chrome.)

**Comment by christophs78** [comment]
Simply doing
`getWebDriver().findElement(By.id("form:fileupload")).sendKeys("C:\\Users\\chris\\IdeaProjects\\wal_jan_primefaces\\primefaces-integration-tests\\target\\test-classes\\org\\primefaces\\integrationtests\\fileupload\\file1.csv") `
works on my machine with Chrome.

Following line seems to be bad for file-input on Chrome:
image

Maybe we should do sendKeys straight forward for this and not use ComponentUtils.

**Comment by wal-jan** [comment]
In principal the tests do now work with Chrome and Mojarra. But now there is timimg issue. Looks like `waitAdvancedUntilAllFilesAreUploaded` does not wait long enough in all cases. I'll have a look on that...

File: primefaces-selenium/primefaces-selenium-components/src/main/java/org/primefaces/selenium/component/FileUpload.java (hunk 2 above)
Comment (christophs78) [reply primefaces-selenium/primefaces-selenium-components/src/main/java/org/primefaces/selenium/component/FileUpload.java:117]: One idea may be using https://github.com/lightbody/browsermob-proxy#http-request-manipulation to monitor whether there is still a active request.

**Review by melloware — APPROVED** [review]

**Comment by melloware** [comment]
Nice everything is passing now!   This is a huge addition to the test suite thank you very much!

File: primefaces-selenium/primefaces-selenium-components/src/main/java/org/primefaces/selenium/component/FileUpload.java (hunk 2 above)
Comment (wal-jan) [reply primefaces-selenium/primefaces-selenium-components/src/main/java/org/primefaces/selenium/component/FileUpload.java:117]: @christophs78 looks promising. I suggest to evaluate this in a separate step (new issue?)

**Comment by wal-jan** [comment]
@melloware There is one aspect, which is not covered in the file upload tests now.
[The SafariDriver does not support file uploads](https://github.com/seleniumhq/selenium-google-code-issue-archive/issues/4220#issue-138332860). I found `Chips004Test#testClipboardPaste(Page)` where Safari is skipped programmatically. I wonder how to define a Test Tag which *excludes* some Tests. For what I know is that `group`  *includes* tagged tests.

**Comment by melloware** [comment]
@wal-jan in InputText001Test we have this...
```java
@Tag("SafariBasic") //example-tag used together with profile/properties/groups in pom.xml to run only tests with this tag
public class InputText001Test extends AbstractPrimePageTest {
```

**Comment by christophs78** [comment]
> @wal-jan in InputText001Test we have this...
> 
> ```java
> @Tag("SafariBasic") //example-tag used together with profile/properties/groups in pom.xml to run only tests with this tag
> public class InputText001Test extends AbstractPrimePageTest {
> ```

Yes, we have this as an example. But it´s not further used for now.

File: primefaces-selenium/primefaces-selenium-components/src/main/java/org/primefaces/selenium/component/FileUpload.java (hunk 2 above)
Comment (christophs78) [reply primefaces-selenium/primefaces-selenium-components/src/main/java/org/primefaces/selenium/component/FileUpload.java:117]: Yes. Opening a new issue for this is a good idea. As always way more available options than available time. ;-) Let´s look what the next weeks/months bring. :-)
Overall i´m really happy with the progress we make on the integration-test-side. A huge thankyou for your contribution. It´s IMO the only way to provide a stable product and allow for continues safe refactoring.

**Comment by melloware** [comment]
Safari also failed all the tests last night...
```
Error:  Errors: 
Error:    FileUpload000Test.testBasicSingleUploadTwice:66 » NoSuchElement 
For documenta...
Error:    FileUpload000Test.testBasicSingleUpload:46 » NoSuchElement 
For documentation ...
Error:    FileUpload001Test.testBasicSingleUploadTwice:70 » ElementNotInteractable The t...
Error:    FileUpload001Test.testBasicSingleUpload:50 » ElementNotInteractable The target...
Error:    FileUpload002Test.testBasicMultipleUploadMultipleFilesTwice:95 » ElementNotInteractable
Error:    FileUpload002Test.testBasicMultipleUploadMultipleFiles:71 » ElementNotInteractable
Error:    FileUpload002Test.testBasicMultipleUploadSingleFile:50 » ElementNotInteractable
Error:    FileUpload003Test.testBasicAutoSingleUploadTwice:68 » ElementNotInteractable T...
Error:    FileUpload003Test.testBasicAutoSingleUpload:50 » ElementNotInteractable The ta...
Error:    FileUpload004Test.testBasicAutoMultipleUploadMultipleFilesTwice:88 » ElementNotInteractable
Error:    FileUpload004Test.testBasicAutoMultipleUploadMultipleFiles:69 » ElementNotInteractable
Error:    FileUpload004Test.testBasicAutoMultipleUploadSingleFile:50 » ElementNotInteractable
Error:    FileUpload005Test.testAdvancedSingleUploadCancel:102 » ElementNotInteractable ...
Error:    FileUpload005Test.testAdvancedSingleUploadTwice:70 » ElementNotInteractable Th...
Error:    FileUpload005Test.testAdvancedSingleUpload:49 » ElementNotInteractable The tar...
Error:    FileUpload006Test.testAdvancedMultipleUploadMultipleFilesTwice:96 » ElementNotInteractable
Error:    FileUpload006Test.testAdvancedMultipleUploadMultipleFiles:71 » ElementNotInteractable
Error:    FileUpload006Test.testAdvancedMultipleUploadSingleFile:49 » ElementNotInteractable
Error:    FileUpload007Test.testAdvancedAutoSingleUploadTwice:68 » ElementNotInteractable
Error:    FileUpload007Test.testAdvancedAutoSingleUpload:49 » ElementNotInteractable The...
Error:    FileUpload008Test.testAdvancedAutoMultipleUploadMultipleFilesTwice:89 » ElementNotInteractable
Error:    FileUpload008Test.testAdvancedAutoMultipleUploadMultipleFiles:69 » ElementNotInteractable
Error:    FileUpload008Test.testAdvancedAutoMultipleUploadSingleFile:49 » ElementNotInteractable
Error:    FileUpload009Test.testAdvancedMultipleUploadMultipleFilesTwice:96 » ElementNotInteractable
Error:    FileUpload009Test.testAdvancedMultipleUploadMultipleFiles:71 » ElementNotInteractable
Error:    FileUpload009Test.testAdvancedMultipleUploadSingleFile:49 » ElementNotInteractable
Error:    FileUpload010Test.testAdvancedAutoMultipleUploadMultipleFilesTwice:89 » ElementNotInteractable
Error:    FileUpload010Test.testAdvancedAutoMultipleUploadMultipleFiles:69 » ElementNotInteractable
Error:    FileUpload010Test.testAdvancedAutoMultipleUploadSingleFile:49 » ElementNotInteractable
```

**Comment by wal-jan** [comment]
@christophs78 @melloware I suggest the Tag `SafariExclude`. If thats OK I'll create a PR with all FileUpload test annotated (and Chips004Test#testClipboardPaste) and `profile[id=Safari]/properties/excludedGroups`

**Comment by melloware** [comment]
absolutely ok with me!

**Pull request merged**